lightly/loss/vicreg_loss.py

```diff
@@ -0,0 +1,98 @@
+import torch
+import torch.distributed as dist
+import torch.nn.functional as F
+
+class VICRegLoss(torch.nn.Module):
+    """Implementation of the VICReg Loss from VICReg[0] paper.
+    This implementation follows the code published by the authors. [1]
+
+    [0] Bardes, A. et. al, 2022, VICReg... https://arxiv.org/abs/2105.04906
+    [1] https://github.com/facebookresearch/vicreg/
+        Examples:
+        >>> # initialize loss function
+        >>> loss_fn = VICRegLoss()
+        >>>
+        >>> # generate two random transforms of images
+        >>> t0 = transforms(images)
+        >>> t1 = transforms(images)
+        >>>
+        >>> # feed through model
+        >>> out0, out1 = model(t0, t1)
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(out0, out1)
+    """
+
+    def __init__(
+        self,
+        lambda_param: float = 25.0,
+        mu_param: float = 25.0,
+        nu_param: float = 1.0,
+        gather_distributed : bool = False,
+        eps = 0.0001
+    ):
+        """Lambda, mu and nu params configuration with default value like in [0]
+        Args:
+            lambda_param:
+                Coefficient for the invariance term of the loss
+                Defaults to 25.0 [0].
+            mu_param:
+                Coefficient for the variance term of the loss
+                Defaults to 25.0 [0].
+            nu_param:
+                Coefficient for the covariance term of the loss
+                Defaults to 1.0 [0].
+            gather_distributed:
+                If True then the cross-correlation matrices from all gpus are
+                gathered and summed before the loss calculation.
+            eps:
+                Numerical epsilon
+                Defaults to 0.0001 [1].
+        """
+        super(VICRegLoss, self).__init__()
+        self.lambda_param = lambda_param
+        self.mu_param = mu_param
+        self.nu_param = nu_param
+        self.gather_distributed = gather_distributed
+
+        self.eps = eps
+
+    def forward(self, z_a: torch.Tensor, z_b: torch.Tensor) -> torch.Tensor:
+
+        # invariance term of the loss
+        repr_loss = F.mse_loss(z_a, z_b)
+
+        # gather all batches
+        if self.gather_distributed and dist.is_initialized():
+            world_size = dist.get_world_size()
+            if world_size > 1:
+                dist.all_reduce(z_a)
```

Comment:
I believe this should be `all_gather` rather than `all_reduce`. What do you think @b-albar  and @guarin?

Reply:
Yes I agree it should be all_gather. But in this case, I think it should use the GatherLayer in lightly.utils.dist to avoid breaking the gradient flow during backprop. I'll fix the PR.